### PR TITLE
[WIP] CCCP API server

### PR DESCRIPTION
**Pull request by navidshaikh**
This changeset adds flask implementation of CCP readonly APIs.

 The API server skeleton is generated by swagger <https://swagger.io/>
 OpenAPI implementation framework using swagger YAML API specifications.

 Added Dockerfile for API server at Dockerfiles/ccp-api-server/Dockerfile,
 which will containerize the API server and start it.

### CODE REVIEW

File: Dockerfiles/ccp-api-server/Dockerfile [hunk 1]
```diff
@@ -17,10 +17,14 @@ ENV PYTHONPATH="/opt/ccp_api_server"
 
 COPY . /opt/ccp_api_server
 
+RUN useradd 1001 && usermod -aG 0 1001 && chown 1001:0 /opt/ccp_api_server && chmod -R g+x /opt/ccp_api_server
```
Comment (navidshaikh) [inline Dockerfiles/ccp-api-server/Dockerfile]: `chown -R 1001:0 /opt/ccp_api_server`

We need to change the permissions to root and all subsequent dirs recursively in that repo, WDYT ?
@mohammedzee1000

File: Dockerfiles/ccp-api-server/Dockerfile [hunk 2]
```diff
@@ -17,10 +17,14 @@ ENV PYTHONPATH="/opt/ccp_api_server"
 
 COPY . /opt/ccp_api_server
```
Comment (navidshaikh) [inline Dockerfiles/ccp-api-server/Dockerfile]: I am thinking, now that `/opt/ccp_api_server` hosts the complete service code, we should name it as our service repo name or `ccp-openshift` as we did in slave container.

File: ccp/apis/server/ccp_server/models/__init__.py [hunk 3]
```diff
@@ -3,23 +3,29 @@
...
-from ccp_server.models.projects import Projects
-from ccp_server.models.scanner_logs import ScannerLogs
-from ccp_server.models.status import Status
-from ccp_server.models.target_file import TargetFile
-from ccp_server.models.weekly_scan_builds_info import WeeklyScanBuildsInfo
-from ccp_server.models.weekly_scan_logs import WeeklyScanLogs
+# from ccp.apis.server.ccp_server.models.all_scanner_logs import AllScannerLogs
+# from ccp.apis.server.ccp_server.models.app_id_job_id_tag import AppIdJobIdTag
+# from ccp.apis.server.ccp_server.models.app_id_job_id_tags import AppIdJobIdTags
+# from ccp.apis.server.ccp_server.models.build_logs import BuildLogs
```
[the remark this answers is not in the document]
Comment (mohammedzee1000) [reply ccp/apis/server/ccp_server/models/__init__.py]: They were causing error, once i updated the references. The errors failed once i removed these imports and the server works fine.

Need to still figure out what happened there

File: Dockerfiles/ccp-api-server/Dockerfile (hunk 1 above)
Comment (mohammedzee1000) [reply Dockerfiles/ccp-api-server/Dockerfile]: This should not be a problem, but to be safe, yea, i guess it makes sense

File: Dockerfiles/ccp-api-server/Dockerfile (hunk 2 above)
Comment (mohammedzee1000) [reply Dockerfiles/ccp-api-server/Dockerfile]: It did right from the start

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml [hunk 4]
```diff
@@ -0,0 +1,671 @@
...
+    example:
+      description: "description"
+      logs: "logs"
+  AllScannerLogs:
+    properties:
+      scanner_name:
+        $ref: "#/definitions/ScannerLogs"
+    title: "All scanner logs"
+    description: "All scanner logs"
+    example:
```
Comment (dharmit) [inline ccp/apis/v1/ccp_server/swagger/swagger.yaml]: There should've been more than one scanners listed for the sake of example.

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml [hunk 5]
```diff
@@ -0,0 +1,671 @@
...
+      scanner_name:
```
Comment (dharmit) [inline ccp/apis/v1/ccp_server/swagger/swagger.yaml]: Is the `scanner_name` going to be replaced with respective scanner names?

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml [hunk 6]
```diff
@@ -0,0 +1,671 @@
...
+        description: "desired-tag of the project"
+        required: true
+        type: "string"
+      responses:
+        200:
+          description: "OK"
+          schema:
+            $ref: "#/definitions/TargetFile"
+      x-swagger-router-controller: "ccp_server.controllers.projects_controller"
+  /namespace/{namespace}/app-id/{app_id}/job-id/{job_id}/desired-tag:
```
[the remark this answers is not in the document]
Comment (bamachrn) [reply ccp/apis/v1/ccp_server/swagger/swagger.yaml]: yes, should not we add `builds` also here? how does the server know that we are asking for builds? like the `target-file ` one, this should be `/namespace/{namespace}/app-id/{app_id}/job-id/{job_id}/desired-tag/{desired_tag}/builds`

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml (hunk 5 above)
Comment (navidshaikh) [reply ccp/apis/v1/ccp_server/swagger/swagger.yaml]: yes, the word `scanner_name` won't make any sense anyway!!

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml (hunk 4 above)
Comment (navidshaikh) [reply ccp/apis/v1/ccp_server/swagger/swagger.yaml]: well!

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/apis/v1/ccp_server/swagger/swagger.yaml]: there was a typo here, fixed it.

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml [hunk 7]
```diff
@@ -0,0 +1,671 @@
...
+        200:
+          description: "All weekly scan builds for given project"
+          schema:
+            $ref: "#/definitions/WeeklyScanBuildsInfo"
+      x-swagger-router-controller: "ccp_server.controllers.builds_controller"
+  /namespace/{namespace}/app-id/{app_id}/job-id/{job_id}/desired-tag/{desired_tag}/build/{build}/logs:
+    get:
+      tags:
+      - "builds"
+      summary: "Build logs for given build number"
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/apis/v1/ccp_server/swagger/swagger.yaml]: this is resolved after review discussions

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml [hunk 8]
```diff
@@ -0,0 +1,671 @@
...
+      build_number:
+        build: "build"
+        status: "status"
+  ProjectWeeklyScanBuilds:
+    properties:
+      build_number:
+        $ref: "#/definitions/ProjectWeeklyScanBuildNameStatus"
+    example:
+      build_number:
+        weeklyscan-build: "weeklyscan-build"
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/apis/v1/ccp_server/swagger/swagger.yaml]: the numbers are going to be a line above that, where `build_number` is referenced.

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml [hunk 9]
```diff
@@ -0,0 +1,671 @@
...
+      produces:
+      - "application/json"
+      parameters: []
+      responses:
+        200:
+          description: "All the readable namespaces"
+          schema:
+            $ref: "#/definitions/Namespaces"
+      x-swagger-router-controller: "ccp_server.controllers.meta_controller"
+  /namespace/{namespace}/projects:
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/apis/v1/ccp_server/swagger/swagger.yaml]: made the respective changes

File: ccp/apis/v1/server/ccp_server/swagger/swagger.yaml [hunk 10]
```diff
@@ -0,0 +1,671 @@
...
+      git_path: "git_path"
+      build_context: "build_context"
+      prebuild_script: "prebuild_script"
+      git_url: "git_url"
+      app_id: "app_id"
+  AppIdJobIdTag:
+    properties:
+      desired_tag:
+        type: "string"
+      build_status:
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/apis/v1/server/ccp_server/swagger/swagger.yaml]: @bamachrn wanted to show the build status infront of desired-tags for a project, thus we added it.

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml [hunk 11]
```diff
@@ -0,0 +1,671 @@
...
+        required: true
+        type: "string"
+      - name: "desired_tag"
+        in: "path"
+        description: "desired-tag of the project"
+        required: true
+        type: "string"
+      responses:
+        200:
+          description: "Al the builds for given project"
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/apis/v1/ccp_server/swagger/swagger.yaml]: fixed

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml [hunk 12]
```diff
@@ -0,0 +1,671 @@
...
+          schema:
+            $ref: "#/definitions/ProjectBuildsInfo"
+      x-swagger-router-controller: "ccp_server.controllers.builds_controller"
+  /namespace/{namespace}/app-id/{app_id}/job-id/{job_id}/desired-tag/{desired_tag}/wscan-builds:
+    get:
+      tags:
+      - "builds"
+      summary: "Get all the weekly scan builds info for given project"
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/apis/v1/ccp_server/swagger/swagger.yaml]: fixed

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml [hunk 13]
```diff
@@ -0,0 +1,671 @@
...
+      description: "Get all the weekly scan builds info for given project"
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/apis/v1/ccp_server/swagger/swagger.yaml]: fixed

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml [hunk 14]
```diff
@@ -0,0 +1,671 @@
...
+        build: "build"
+        status: "status"
+  ProjectWeeklyScanBuilds:
+    properties:
+      build_number:
+        $ref: "#/definitions/ProjectWeeklyScanBuildNameStatus"
+    example:
+      build_number:
+        weeklyscan-build: "weeklyscan-build"
+        status: "status"
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/apis/v1/ccp_server/swagger/swagger.yaml]: fixed

File: ccp/apis/v1/ccp_server/swagger/swagger.yaml (hunk 4 above)
Comment (navidshaikh) [reply ccp/apis/v1/ccp_server/swagger/swagger.yaml]: added examples of two scanners to be explicit!!!!

**Comment by bamachrn** [comment]
merged this in #668

**Pull request closed without merging**